src/nvim/event/process.c

```diff
@@ -293,28 +309,48 @@ static void process_close(Process *proc)
   }
 }
 
-static void on_process_exit(Process *proc)
+static void process_close_out_handles(void **argv)
 {
-  if (exiting) {
-    on_process_exit_event((Event) {.data = proc});
-  } else {
-    loop_push_event(proc->loop,
-        (Event) {.handler = on_process_exit_event, .data = proc}, false);
+  Process *proc = argv[0];
+  process_close_out(proc);
+  process_close_err(proc);
+}
+
+static void eof_timeout_cb(uv_timer_t *handle)
+{
+  Process *proc = handle->data;
+  uv_timer_stop(handle);
+  uv_close((uv_handle_t *)handle, NULL);
+  queue_put(proc->events, process_close_out_handles, 1, proc);
+}
+
+static void process_close_handles(void **argv)
+{
+  Process *proc = argv[0];
+  process_close_in(proc);
+  process_close(proc);
+  if ((proc->out && !proc->out->closed) || (proc->err && !proc->err->closed)) {
+    uv_timer_init(&proc->loop->uv, &proc->eof_timer);
+    proc->eof_timer.data = proc;
+    uv_timer_start(&proc->eof_timer, eof_timeout_cb, proc->stream_eof_timeout,
+        proc->stream_eof_timeout);
   }
+}
 
+static void on_process_exit(Process *proc)
+{
   Loop *loop = proc->loop;
   if (loop->children_stop_requests && !--loop->children_stop_requests) {
```

Comment:
A process that was not stopped (`proc->stopped_time==0`) should not modify `loop->children_stop_requests` and stop the process kill timer. Stopped processes could continue running, because no kil signal will be sent.